lib/src/util/condition_scope_visitor.dart

```diff
@@ -0,0 +1,256 @@
+import 'dart:collection';
+
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/visitor.dart';
+import 'package:analyzer/dart/element/element.dart';
+import 'package:linter/src/util/dart_type_utilities.dart';
+
+Element _getLeftElement(AssignmentExpression assignment) {
+  final leftPart = assignment.leftHandSide;
+  return leftPart is SimpleIdentifier
+      ? DartTypeUtilities.getCanonicalElement(leftPart.bestElement)
+      : leftPart is PropertyAccess
+          ? DartTypeUtilities
+              .getCanonicalElement(leftPart.propertyName.bestElement)
+          : null;
+}
+
+bool _isBreakStatement(Statement statement) => (statement is BreakStatement ||
+    (statement is Block &&
+        statement.statements.length == 1 &&
+        statement.statements.first is BreakStatement));
+
+bool _isContinueStatement(
+        Statement statement) =>
+    (statement is ContinueStatement ||
+        (statement is Block &&
+            statement.statements.length == 1 &&
+            statement.statements.first is ContinueStatement));
+
+bool _isReturnStatement(Statement statement) => (statement is ReturnStatement ||
+    (statement is Block &&
+        statement.statements.length == 1 &&
+        statement.statements.first is ReturnStatement));
+
+abstract class ConditionScopeVisitor extends RecursiveAstVisitor {
+  final Queue<Queue<_ExpressionBox>> environments = new Queue();
+
+  Iterable<Expression> getFalseExpressions(Iterable<Element> elements) =>
+      _getExpressions(elements, value: false);
+
+  Iterable<Expression> getTrueExpressions(Iterable<Element> elements) =>
+      _getExpressions(elements);
+
+  @override
+  visitAssignmentExpression(AssignmentExpression node) {
+    _addElementToEnvironment(new _UndefinedExpression(_getLeftElement(node)));
```

Comment:
Yes it is harmless, but other things like:

```dart
if (notDefined < 2) {
  if (notDefined < 2) { // there is a lint here
  }
}
```
can decrease the UX level, because there will be a linter and an error in the same place, I fixed this locally but it is necessary to check for types and it is really expensive (the benchmark in the rule moved from ~55 to >140), so I think it is a cost I would pay to have a faster linter as a user.

If you think this is OK the branch is ready to merge, if you want me to fix this even that means have a worse performance just tell me and I will do it.